Loader cannot be made into an object

Hi,

I am trying to update this dependency from `0.5.16` to `0.8` and I reference the `Loader` trait in a struct field:

```rust
/// Lookup a language string in the underlying loader.
pub struct FluentHelper {
    loader: Box<dyn Loader + Send + Sync>,
    /// Escape messages, default is `true`.
    pub escape: bool,
}
```

Now I am unable to use the trait boxed due to this compiler error:

```
error[E0038]: the trait `Loader` cannot be made into an object
  --> src/lib.rs:95:17
   |
95 |     loader: Box<dyn Loader + Send + Sync +'static>,
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Loader` cannot be made into an object
   |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  --> /Users/muji/.cargo/registry/src/github.com-1ecc6299db9ec823/fluent-templates-0.8.0/src/loader.rs:35:8
   |
35 |     fn lookup_with_args<T: AsRef<str>>(
   |        ^^^^^^^^^^^^^^^^ the trait cannot be made into an object because method `lookup_with_args` has generic type parameters
...
45 |     fn lookup_complete<T: AsRef<str>>(
   |        ^^^^^^^^^^^^^^^ the trait cannot be made into an object because method `lookup_complete` has generic type parameters
```

I think it maybe be ok for `Loader` to just accept `&str`, I wonder what you think?

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -342,3 +342,24 @@ pub use once_cell;
 
 /// A convenience `Result` type that defaults to `error::Loader`.
 pub type Result<T, E = error::LoaderError> = std::result::Result<T, E>;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn check_if_loader_is_object_safe() {
+        use crate::Loader;
+        use unic_langid::{langid, LanguageIdentifier};
+
+        const US_ENGLISH: LanguageIdentifier = langid!("en-US");
+
+        let loader = ArcLoader::builder("./tests/locales", US_ENGLISH)
+            .customize(|bundle| bundle.set_use_isolating(false))
+            .build()
+            .unwrap();
+
+        let loader: Box<dyn Loader> = Box::new(loader);
+        assert_eq!("Hello World!", loader.lookup(&US_ENGLISH, "hello-world"));
+    }
+}

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -160,7 +160,7 @@
 //!
 //!     let args = {
 //!         let mut map = HashMap::new();
-//!         map.insert(String::from("name"), "Alice".into());
+//!         map.insert(Cow::from("name"), "Alice".into());
 //!         map
 //!     };
 //!
@@ -333,9 +333,9 @@ pub mod loader;
 
 #[cfg(feature = "macros")]
 pub use fluent_template_macros::static_loader;
-pub use unic_langid::LanguageIdentifier;
 #[cfg(feature = "macros")]
 pub use unic_langid::langid;
+pub use unic_langid::LanguageIdentifier;
 
 #[doc(hidden)]
 pub use once_cell;

--- a/src/loader.rs
+++ b/src/loader.rs
@@ -10,6 +10,7 @@ mod tera;
 
 mod shared;
 
+use std::borrow::Cow;
 use std::collections::HashMap;
 
 use crate::FluentBundle;
@@ -28,48 +29,48 @@ pub use static_loader::StaticLoader;
 pub trait Loader {
     /// Look up `text_id` for `lang` in Fluent.
     fn lookup(&self, lang: &LanguageIdentifier, text_id: &str) -> String {
-        self.lookup_complete::<&str>(lang, text_id, None)
+        self.lookup_complete(lang, text_id, None)
     }
 
     /// Look up `text_id` for `lang` with `args` in Fluent.
-    fn lookup_with_args<T: AsRef<str>>(
+    fn lookup_with_args(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: &HashMap<T, FluentValue>,
+        args: &HashMap<Cow<'static, str>, FluentValue>,
     ) -> String {
         self.lookup_complete(lang, text_id, Some(args))
     }
 
     /// Look up `text_id` for `lang` in Fluent, using any `args` if provided.
-    fn lookup_complete<T: AsRef<str>>(
+    fn lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> String;
 
     /// Look up `text_id` for `lang` in Fluent.
     fn try_lookup(&self, lang: &LanguageIdentifier, text_id: &str) -> Option<String> {
-        self.try_lookup_complete::<&str>(lang, text_id, None)
+        self.try_lookup_complete(lang, text_id, None)
     }
 
     /// Look up `text_id` for `lang` with `args` in Fluent.
-    fn try_lookup_with_args<T: AsRef<str>>(
+    fn try_lookup_with_args(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: &HashMap<T, FluentValue>,
+        args: &HashMap<Cow<'static, str>, FluentValue>,
     ) -> Option<String> {
         self.try_lookup_complete(lang, text_id, Some(args))
     }
 
     /// Look up `text_id` for `lang` in Fluent, using any `args` if provided.
-    fn try_lookup_complete<T: AsRef<str>>(
+    fn try_lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> Option<String>;
 
     /// Returns an Iterator over the locales that are present.
@@ -80,20 +81,20 @@ impl<L> Loader for std::sync::Arc<L>
 where
     L: Loader,
 {
-    fn lookup_complete<T: AsRef<str>>(
+    fn lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> String {
         L::lookup_complete(self, lang, text_id, args)
     }
 
-    fn try_lookup_complete<T: AsRef<str>>(
+    fn try_lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> Option<String> {
         L::try_lookup_complete(self, lang, text_id, args)
     }
@@ -107,20 +108,20 @@ impl<'a, L> Loader for &'a L
 where
     L: Loader,
 {
-    fn lookup_complete<T: AsRef<str>>(
+    fn lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> String {
         L::lookup_complete(self, lang, text_id, args)
     }
 
-    fn try_lookup_complete<T: AsRef<str>>(
+    fn try_lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> Option<String> {
         L::try_lookup_complete(self, lang, text_id, args)
     }

--- a/src/loader/arc_loader.rs
+++ b/src/loader/arc_loader.rs
@@ -1,3 +1,4 @@
+use std::borrow::Cow;
 use std::collections::HashMap;
 use std::fs::read_dir;
 use std::path::{Path, PathBuf};
@@ -107,11 +108,11 @@ pub struct ArcLoader {
 
 impl super::Loader for ArcLoader {
     // Traverse the fallback chain,
-    fn lookup_complete<T: AsRef<str>>(
+    fn lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> String {
         for lang in negotiate_languages(&[lang], &self.bundles.keys().collect::<Vec<_>>(), None) {
             if let Ok(val) = self.lookup_single_language(lang, text_id, args) {
@@ -127,11 +128,11 @@ impl super::Loader for ArcLoader {
     }
 
     // Traverse the fallback chain,
-    fn try_lookup_complete<T: AsRef<str>>(
+    fn try_lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> Option<String> {
         for lang in negotiate_languages(&[lang], &self.bundles.keys().collect::<Vec<_>>(), None) {
             if let Ok(val) = self.lookup_single_language(lang, text_id, args) {

--- a/src/loader/handlebars.rs
+++ b/src/loader/handlebars.rs
@@ -6,6 +6,7 @@ use handlebars::{
 use fluent_bundle::FluentValue;
 use handlebars::template::{Parameter, TemplateElement};
 use serde_json::Value as Json;
+use std::borrow::Cow;
 use std::collections::HashMap;
 
 use crate::{FluentLoader, Loader};
@@ -57,7 +58,7 @@ impl<L: Loader + Send + Sync> HelperDef for FluentLoader<L> {
             .into());
         };
 
-        let mut args: Option<HashMap<String, FluentValue>> = if h.hash().is_empty() {
+        let mut args: Option<HashMap<Cow<'static, str>, FluentValue>> = if h.hash().is_empty() {
             None
         } else {
             let map = h
@@ -73,7 +74,7 @@ impl<L: Loader + Send + Sync> HelperDef for FluentLoader<L> {
                         Json::String(s) => s.to_owned().into(),
                         _ => return None,
                     };
-                    Some((k.to_string(), val))
+                    Some((Cow::from(k.to_string()), val))
                 })
                 .collect();
             Some(map)
@@ -112,7 +113,10 @@ impl<L: Loader + Send + Sync> HelperDef for FluentLoader<L> {
                     if let Some(ref tpl) = block.template {
                         let mut s = StringOutput::default();
                         tpl.render(reg, context, rcx, &mut s)?;
-                        args.insert(String::from(id), FluentValue::String(s.s.into()));
+                        args.insert(
+                            Cow::Owned(String::from(id)),
+                            FluentValue::String(s.s.into()),
+                        );
                     }
                 }
             }

--- a/src/loader/static_loader.rs
+++ b/src/loader/static_loader.rs
@@ -1,3 +1,4 @@
+use std::borrow::Cow;
 use std::collections::HashMap;
 
 use crate::{error::LookupError, languages::negotiate_languages, FluentBundle};
@@ -62,11 +63,11 @@ impl StaticLoader {
 
 impl super::Loader for StaticLoader {
     // Traverse the fallback chain,
-    fn lookup_complete<T: AsRef<str>>(
+    fn lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> String {
         for lang in negotiate_languages(&[lang], &self.bundles.keys().collect::<Vec<_>>(), None) {
             if let Ok(val) = self.lookup_single_language(lang, text_id, args) {
@@ -83,11 +84,11 @@ impl super::Loader for StaticLoader {
     }
 
     // Traverse the fallback chain,
-    fn try_lookup_complete<T: AsRef<str>>(
+    fn try_lookup_complete(
         &self,
         lang: &LanguageIdentifier,
         text_id: &str,
-        args: Option<&HashMap<T, FluentValue>>,
+        args: Option<&HashMap<Cow<'static, str>, FluentValue>>,
     ) -> Option<String> {
         for lang in negotiate_languages(&[lang], &self.bundles.keys().collect::<Vec<_>>(), None) {
             if let Ok(val) = self.lookup_single_language(lang, text_id, args) {

--- a/src/loader/tera.rs
+++ b/src/loader/tera.rs
@@ -1,5 +1,6 @@
 use fluent_bundle::FluentValue;
 use serde_json::Value as Json;
+use std::borrow::Cow;
 use std::collections::HashMap;
 use unic_langid::LanguageIdentifier;
 
@@ -66,7 +67,7 @@ impl<L: Loader + Send + Sync> tera::Function for crate::FluentLoader<L> {
 
         for (key, value) in args.iter().filter(is_not_tera_key) {
             fluent_args.insert(
-                heck::ToKebabCase::to_kebab_case(&**key),
+                Cow::from(heck::ToKebabCase::to_kebab_case(&**key)),
                 json_to_fluent(value.clone())?,
             );
         }

